### PR TITLE
Make X-Ray fully compatible with Sodium

### PR DESCRIPTION
**Pull request by ZANX3Y**
## Description
Made X-Ray fully compatible with Sodium.
Liquids, plants, etc. now work the same way as in Vanilla.
Completes #458

### CODE REVIEW

File: src/main/java/net/wurstclient/mixin/SodiumBlockRendererMixin.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    private void getGeometry(BlockRenderContext ctx, Direction face,
+		CallbackInfoReturnable<List<BakedQuad>> cir)
+    {
+        if (face != null) return;
+
+        ShouldDrawSideEvent event = new ShouldDrawSideEvent(ctx.state(), ctx.pos());
+        EventManager.fire(event);
+
+        if (event.isRendered() != null && !event.isRendered())
+            cir.setReturnValue(List.of());
```
[the remark this answers is not in the document]
Comment (ZANX3Y) [reply src/main/java/net/wurstclient/mixin/SodiumBlockRendererMixin.java:40]: The `isRendered` method is just a getter.

**Review by Alexander01998 — CHANGES_REQUESTED** [review]
Hey @ZANX3Y, great work on making X-Ray compatible with Sodium! 👍

However, I'm worried that adding Sodium as a dependency might cause problems with future updates, since Sodium doesn't support Minecraft's snapshots and often takes a while to support new releases.

To address this, could you modify the mixins to work without the Sodium dependency, similar to how `SodiumBlockOcclusionCacheMixin` worked previously? This way, we can maintain compatibility with Sodium without directly linking our update cycle to theirs.

Thanks!

**Comment by ZANX3Y** [comment]
Done. Had to create 3 BakedModel mixins to replace the BlockRenderer mixin, but now it works without having Sodium as a dependency.

File: src/main/java/net/wurstclient/hacks/XRayHack.java [hunk 2]
```diff
@@ -21,12 +21,7 @@
 import net.wurstclient.Category;
 import net.wurstclient.SearchTags;
 import net.wurstclient.clickgui.screens.EditBlockListScreen;
-import net.wurstclient.events.GetAmbientOcclusionLightLevelListener;
-import net.wurstclient.events.RenderBlockEntityListener;
-import net.wurstclient.events.SetOpaqueCubeListener;
-import net.wurstclient.events.ShouldDrawSideListener;
-import net.wurstclient.events.TesselateBlockListener;
-import net.wurstclient.events.UpdateListener;
+import net.wurstclient.events.*;
```
[the remark this answers is not in the document]
Comment (Alexander01998) [reply src/main/java/net/wurstclient/hacks/XRayHack.java]: This is fine. @nobotreply

**Comment by Alexander01998** [comment]
Hey @ZANX3Y, thanks for the update!

I see why you had to replace the `BlockRenderer` mixin - the `BlockRenderContext` parameter is not accessible without having Sodium as a dependency.

But after doing some testing, it seems we might not actually need all 3 of those `BakedModel` mixins. Specifically, `MultipartBakedModelMixin` and `WeightedBakedModelMixin` don't seem to have any effect on X-Ray's rendering, unless I missed something? All of these blocks still get X-Ray'd correctly with those two mixins removed:

A screenshot showing various blocks with X-Ray disabled.

A screenshot from the same angle but with X-Ray enabled. All of the various blocks have turned invisible.

Please let me know if you've run into any blocks or scenarios that don't work properly without these mixins. Otherwise, let's remove them to simplify the logic.

Also, I found that `BasicBakedModelMixin` actually works even when Sodium isn't installed. If we remove the stack trace check that filters out non-Sodium method calls, this mixin can replace `TerrainRenderContextMixin` and thereby further simplify the X-Ray code.

In theory, this would even make `BlockModelRendererMixin` redundant. This mixin does pretty much the same things as `TerrainRenderContextMixin`, but only when OptiFine is installed. Since OptiFine is not yet available for Minecraft 1.20.2, I can't 100% confirm this right now.

I'll await your response before making these changes, just in case I got anything wrong here.

**Comment by ZANX3Y** [comment]
Hi @Alexander01998, looks like you're right.

I flew around the "Debug mode" world and confirmed that both `MultipartBakedModelMixin` and `WeightedBakedModelMixin` are actually not needed.

I also tried removing the stack trace check and `TerrainRenderContextMixin` as you suggested. It seems that `BasicBakedModelMixin` handles everything well for Sodium and vanilla renderer.

As for the `BlockModelRendererMixin`, I suspect that it should be redundant, unless OptiFine does some sort of caching of the model quads.

Cheers

**Review by Alexander01998 — APPROVED** [review]

**Pull request merged**